### ISSUE TITLE
Option to avoid splitting statements

### ISSUE THREAD

**Issue by dcsobral** [mvdan/sh#250]
Add an option to avoid splitting statements separated by semicolons into multiple lines.

**Comment by alok** [comment]
I think this, and most of the other suggested changes, should not be added since they defeat the point of an autoformatter.

**Comment by mvdan** [comment]
I agree with @alok - this is a pretty basic feature of the formatter. It also seems like some issues, like #248, overlap with this one.

Note that the formatter already leaves certain one-liners alone, like `if foo; then bar; fi`. If you have specific examples of code that makes sense to leave as a single line, please do share them. I'm going to leave this issue open for a few days, just in case.

**Comment by dcsobral** [comment]
I had one case where the line splitting annoyed me, and, yes, if I could
turn off formatting for a few lines it would be fine too.

I had a series of statements like

VAR=VALUE;(spacing)export VAR

I'll grant that I could make then "export VAR=VALUE", but I liked them the
way they were, and they were all split. If I used the option to preserve
horizontal spacing, it would split and place the export statement on the
same column it was before.

The thing is, if I put two statements on the same line, it is always by
design, not chance, so, for me, if the tool is splitting statements, it's
undoing something I wanted. From a "let's unify everyone's style" PoV,
that's fine. From a "help me prettify my script" PoV, it's not.

As to "defeating the point of an auto-formatter", I'm used to highly
configurable auto-formatters, What seems an eminently reasonable choice to
some people are nonsensical to others, and I've never seen a team switch
format to what a tool implements, and, always, instead, adopt the tool if
it can be configured for their style. It is in that vein that I made all
the tickets: things the tool was doing that would not be accepted by some
teams I've been part of.

**Comment by mvdan** [comment]
> From a "let's unify everyone's style" PoV, that's fine. From a "help me prettify my script" PoV, it's not.

That's a fine balance. If the tool did too little, such as only handle indentation, some people would be angry. If it did everything, like always enforce separate lines, other people would be angry.

> The thing is, if I put two statements on the same line, it is always by design, not chance

That is an interesting point. However, most people would use the tool in the opposite direction - to keep others from having one-liners that are too clever, and to make code more readable.

Do you have examples of large projects that use this kind of style on purpose, or style guides that enforce it? I generally only add options if there's evidence that they'd satisfy lots of users. For now, this seems like a very niche style.

> I'm used to highly configurable auto-formatters,

As I've said in the other thread, this tool will not become `indent` with its 100 or so options and knobs. There are six formatting options so far, and I don't intend to go past that number easily :)

A formatter with fifty more knobs and options would be an entirely different project - one that I don't have time for. If someone wants to take that on, they could reuse much of my work - such as the parser and the syntax tree. I simply don't have time to write and maintain such a thing.

**Comment by zwarmapapa** [comment]
I have several cases where this applies that cause this formatter to not be used in our projects (because it turns simple things into a mess because of the statement splitting, which is the very exact opposite of what you want a formatter to do!) 

Some of those cases are:

```
(crontab -l; echo '* * * * *   /bin/bash /etc/myapps/broadcaster/run.sh >/dev/null 2>&1') | crontab -
(crontab -l; echo '*/5 * * * *   /bin/bash /etc/myapps/broadcaster/update.sh >/dev/null 2>&1') | crontab -

# turns into;

(
    crontab -l
    echo '* * * * *   /bin/bash /etc/myapps/broadcaster/run.sh >/dev/null 2>&1'
) | crontab -
(
    crontab -l
    echo '*/5 * * * *   /bin/bash /etc/myapps/broadcaster/update.sh >/dev/null 2>&1'
) | crontab -
```


```
flock -w 600 "$DEPLOYER_LOCK_UPDATE" || { echo "Build is still running..."; exit 0; }
{
    ...
}
flock -u "$DEPLOYER_LOCK_UPDATE"

# turns into:

flock -w 600 "$DEPLOYER_LOCK_UPDATE" || {
    echo "Build is still running..."
    exit 0
}
{
    ...
}
flock -u "$DEPLOYER_LOCK_UPDATE"
```


```
if [ -f 'build.sh' ]; then buildScriptPath='build.sh'
elif [ -f 'bin/build.sh' ]; then buildScriptPath='bin/build.sh'
elif [ -f 'Bin/build.sh' ]; then buildScriptPath='Bin/build.sh'
elif [ -f 'scripts/build.sh' ]; then buildScriptPath='scripts/build.sh'
elif [ -f 'Scripts/build.sh' ]; then buildScriptPath='Scripts/build.sh'
fi

# turns into

if [ -f 'build.sh' ]; then
    buildScriptPath='build.sh'
elif [ -f 'bin/build.sh' ]; then
    buildScriptPath='bin/build.sh'
elif [ -f 'Bin/build.sh' ]; then
    buildScriptPath='Bin/build.sh'
elif [ -f 'scripts/build.sh' ]; then
    buildScriptPath='scripts/build.sh'
elif [ -f 'Scripts/build.sh' ]; then
    buildScriptPath='Scripts/build.sh'
fi
```


I think it is exceptionally pompous to think that one developer knows what's best for every single project out there and thereby enforcing word splitting in every single case. It's an overestimation of your own knowledge which is a result of nothing more than common ignorance. 

Take this as you will, but truth is, you don't even know how many people you've turned away from this formatter with this as most people don't even bother saying a single thing, they just switch to another formatter, or they even opt out of any shell formatting altogether.

**Comment by mvdan** [comment]
Please keep it civil. You can give honest feedback without implying that the developer is being pompous or ignorant. Especially when the developer is doing this in their spare time and publishing the code as open source.

Back to the topic at hand, adding new flags is always a tradeoff. Any formatting flag is extra load to develop and maintain, and it makes the formatter harder to use and less consistent depending on what flags each user chooses. So the bar for more flags is high. All that said, there is https://github.com/mvdan/sh/issues/261, which is my thinking of how such a flag could be added.

If you mean never splitting lines, then I would argue that the formatter would be far too forgiving. The point of a formatter is to find the balance between consistent style and not bothering the developer too much. There has to be some sort of limit to how much one can cram into a single line.

**Comment by zwarmapapa** [comment]
The formatter can turn off all splitting, which would work just fine. Teams have code review / quality assurance themselves which will take care of any case of cramming everything into a single line. 

A simple all-statement-splitting-off option is all that is needed to make it usable. The more formatting style is forced upon the user/teams, the less likely they are to be able to use it. 

It's currently unusable in many teams, the teams I'm currently in aren't able to use it as it has an adverse effect on the quality and readability of their code. 

Calling it pompous would be on point as expecting everyone to adhere to a single unchangeable style would be exactly that. It's not realistic, and it has a negative impact on the popularity of this formatter. A single style will not fit every usecase, every team, every project, it's simply unrealistic. 

I know the downside that options add to the code, adding a bunch of if/else statements creates a mess, but the end user only cares about being able to use the formatter without it making their code worse rather than better. The more options that are added, the more users/teams will be able to use this formatter. 

In the end it's your project, your say, the popularity and the future of this formatter is yours to decide. All we can do is give feedback and hope for the best, and otherwise something else will take its place instead.

**Comment by mvdan** [comment]
Thanks. I appreciate your feedback, and #261 is open. If anyone wants to contribute there or open a pull request, that would be welcome.

I just want to reiterate that I'm not going to respond well to name-calling or to implications that the software is unusable or a failure waiting to be replaced. If anything, it just makes me less likely to spend any more time on the topic.

**Comment by mvdan** [comment]
In case anyone comes to this thread later and wonders why it seems like I was talking to myself last week, it seems like the other user deleted their comments. Annoyingly enough, GitHub doesn't show that anywhere in the UI. Luckily, I took [a snapshot](https://web.archive.org/web/20200619155625/https://github.com/mvdan/sh/issues/250).

**Comment by patrickvane** [comment]
A fork that will never split statements can be found here: https://github.com/patrickvane/sh/blob/patrickvane/shfmt.exe
The option is "-ns" (never split). 
To use it in phpstorm etc, make a cmd like this: https://github.com/patrickvane/sh/blob/patrickvane/shfmt.cmd
